### PR TITLE
Make `—order random` more deterministic.

### PR DESCRIPTION
**Pull request by myronmarston**
When you’re troubleshooting an order dependent
failure, you want to get the repro case down to
a minimal run that loads and runs as few specs
as possible. With the old random ordering implementation,
that was hard to achieve because while rerunning with
a given seed produced the same order when the exact
same set of examples were loaded, the ordering would
be completely different when a subset was loaded.

By ordering by `md5(seed + example_id)` it ensures
that the ordering of any two examples should stay
consistent regardless of how many other examples are
loaded.

MD5 is significantly slower than `shuffle`, and it
loads an additional part of our stdlib (which we
want to minimize) but I think the tradeoff is worth it
here.

This will help us be able to implement the `--bisect` option discussed in #1767.

What do others think?  Is the tradeoff worth it?  Is there some other hashing function we should consider using?  If there was a hashing function that had similar distribution characteristics to MD5 and was faster and easy to use w/o loading part of the stdlib, that would be ideal, but I can't think of one.

/cc @rspec/rspec

### CODE REVIEW

**Comment by samphippen** [comment]
@myronmarston was there a reason you went for md5 instead of the `#hash` of the string?

Using `#hash` instead of md5 has the advantage of not doing a the require, and might well be faster (the hash algorithm is cheaper because it's not cryptographic). The disadvantage is that we'll likely see more hash value collisions. I'm not sure if that's a problem. Thoughts?

**Comment by myronmarston** [comment]
> @myronmarston was there a reason you went for md5 instead of the #hash of the string?
> 
> Using #hash instead of md5 has the advantage of not doing a the require, and might well be faster (the hash algorithm is cheaper because it's not cryptographic). The disadvantage is that we'll likely see more hash value collisions. I'm not sure if that's a problem. Thoughts?

I tried out `hash` but the problem is that it is not deterministic across processes.  `"foo".hash` within a single ruby process returns a deterministic value, but will return a different value in another ruby process:

```
➜  ~  ruby -v
ruby 2.1.5p273 (2014-11-13 revision 48405) [x86_64-darwin12.0]
➜  ~  ruby -e 'puts "foo".hash'
2464062624698253809
➜  ~  ruby -e 'puts "foo".hash'
-1837013944110794440
➜  ~  ruby -e 'puts "foo".hash'
-1155068896409297005
➜  ~  ruby -e 'puts "foo".hash'
-97344533053661111
```

Thus, we can't use `hash`, unfortunately.

Any other ideas for what we can use, @samphippen?  Obviously, "cryptographically secure" doesn't matter here (which MD5 is not, but it was originally intended or that, IIRC), so I imagine there are better hash functions to use than MD5.

**Comment by samphippen** [comment]
@myronmarston in university I implemented this alg: https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function in C, it was pretty legit. Is it worth rolling an impl of some standard hash instead of doing the MD5 require?

**Comment by myronmarston** [comment]
> @myronmarston in university I implemented this alg: https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function in C, it was pretty legit. Is it worth rolling an impl of some standard hash instead of doing the MD5 require?

Maybe.  I don't really want to get in the business of maintaining a hash function (not my area of expertise or something I care to get involved with) but if there's a simple one that's not much code and has the desired properties, I'm open to it.  I worry that it might be hard to beat MD5 speeds since it's written in C, but maybe that one you linked to is simple enough?

**Comment by samphippen** [comment]
It's worth benchmarking. That algorithm is standard and won't change. MD5 has a bunch of cost associated with startup and hash finalization that cheaper string hashing algorithms don't. I'm just mildly worried that MD5 is a common module that people require and potentially hiding an error for them could be a big issue.

**Comment by samphippen** [comment]
(We should use each_byte to deal with unicode strings tho).

**Comment by samphippen** [comment]
@myronmarston if it helps, here's an implementation of FNV: https://gist.github.com/samphippen/3e8a93757cad9af47f5a

**Comment by myronmarston** [comment]
That looks pretty simple. What are the distribution characterstics like?  Can you throw together a benchmark comparing it to MD5?

BTW, is the initial hash value simply a value that's known to work well for the algo?

> I'm just mildly worried that MD5 is a common module that people require and potentially hiding an error for them could be a big issue.

We try to minimize stdlib requires as a courtesy to the user to minimize how much RSpec mutates the global namespace but I would never consider it a "big issue" to load a std lib piece.  It's something to weigh against the benefits we gain from using a particular stdlib (and I'm glad we do that), but I think we perhaps worry more than we should.

I like the idea of using this hash function you've proposed, though.

**Comment by soulcutter** [comment]
I use a Jenkins hash in one of my projects for work, it's pretty fast

**Comment by soulcutter** [comment]
Looking at FNV it's probably faster than Jenkins - there's not many operations in there

**Comment by myronmarston** [comment]
@soulcutter -- got a link for jenkins hash?  I've never heard of it.

**Comment by soulcutter** [comment]
http://en.wikipedia.org/wiki/Jenkins_hash_function

https://gist.github.com/soulcutter/4c55548bd742ff26a9a0

It looks like the ruby implementation we've got is doing a lot more to try to bound the number. I wonder if that's necessary…

**Comment by JonRowe** [comment]
:+1: for using @samphippen's version, he does have a reasonable crypto background too in terms of maintaining it!

**Comment by soulcutter** [comment]
Yeah, after reading a bunch of hash wikipedia articles I think FNV's a good choice and the implementation is very reasonable. Benchmarking seems prudent, but I'm relatively certain it's going to beat md5.

**Comment by JonRowe** [comment]
Even if it didn't it'd be worth having to avoid the std lib require

**Comment by myronmarston** [comment]
Cool.  I'll try pulling in @samphippen's FNV implementation later today or tomorrow.

File: spec/rspec/core/ordering_spec.rb [hunk 1]
```diff
@@ -26,6 +30,37 @@ module Ordering
...
+
+          it 'has a good distribution', :slow do
+            orderings = 1.upto(1000).map do |seed|
+              order_with(seed)
+            end.uniq
+
+            # Here we are making sure that using MD5 for ordering has a good distribution.
+            # Each seed produces a deterministic order and we want 99%+ of 1000 to be
+            # different.
+            expect(orderings.count).to be > 990
```
Comment (xaviershay) [inline spec/rspec/core/ordering_spec.rb]: what's the expected probability of this failing? :P

**Comment by xaviershay** [comment]
this is a neat idea

**Comment by myronmarston** [comment]
So either @samphippen's implementation of FNV is wrong or FNV's distribution characteristics are _awful_ or I'm doing something wrong I don't see.

``` ruby
require 'digest/md5'

def fowler_noll_vo_hash(string)
  hash = 0xcbf29ce484222325
  string.each_byte do |byte|
    hash = hash * 0x100000001b3
    hash = hash & ((1 << 64)-1)
    hash = hash ^ byte
    hash = hash & ((1 << 64)-1)
  end

  hash
end

def id_from(n)
  "./some_spec.rb[1:#{n}]"
end

def unique_orderings(hash_function)
  list = 10.times.map { |n| id_from(n) }
  1.upto(1000).map do |seed|
    list.sort_by { |id| hash_function.call(seed.to_s + id) }
  end.uniq.count
end

puts "Unique orderings for MD5: #{unique_orderings(Digest::MD5.method(:digest))}"
puts "Unique orderings for FNV: #{unique_orderings(method(:fowler_noll_vo_hash))}"
```

Produces:

```
Unique orderings for MD5: 1000
Unique orderings for FNV: 16
```

Sorting a list of 10 items 1000 times produced 1000 different orderings for MD5 and only 16 for FNV.  Any idea what's up with that result?

**Comment by myronmarston** [comment]
OK, I pushed some updates. Notes:

* FNV doesn't have a good distribution, as I noted above, but jenkins appears to.  I sorted a list of 10 items using 50K different seeds and got 49,664 unique orderings (which beats MD5 -- it got 49,638!).
* While an apples-to-apples comparison of Jenkins and MD5 may show Jenkins is faster (as the benchmark @soulcutter linked to shows), we're comparing Jenkins-in-Ruby vs MD5-in-C, and `Digest::MD5` is still significantly faster (about 2x) than Jenkins.
* The speed of the hash function looks like it actually doesn't matter much; in my sorting benchmarks, MD5 vs Jenkins are head-to-head and are both dwarfed by `shuffle`.  It appears that the main difference is `sort_by` vs `shuffle`, and the actual thing you sort by matters much less...which makes a lot of sense if you think about it...sorting has to do the `O(N log N)` work of sorting, regardless of what the `sort_by` block does.  `shuffle`, OTOH, doesn't need to do the work of sorting -- I suspect it's linear (I haven't looked at the algorithm, though).
* 1.8.7 is failing and will continue to do so until rspec/rspec-expectations#758 is merged.

**Comment by soulcutter** [comment]
That makes sense regarding C md5 vs ruby jenkins. It's too bad we can't use a C jenkins extension!

I feel as though avoiding the stdlib require may be a good thing if performance is similar. I could go either way, though, because I understand not necessarily wanting to maintain a hash function implementation that isn't really vetted.

I can say that I had this implementation matching the output of a different C implementation, so it's probably mostly ok. The main thing that probably matters is that it's consistent and it has a good distribution, right?

I suppose I lean slightly towards this jenkins implementation. What do you think @samphippen ?

**Comment by myronmarston** [comment]
So this is green and it's ready to merge as far as I'm concerned.  @samphippen, do you want to apply your knowledge in this area and review our use of Jenkins before I merge? (Or anyone, really, I just mentioned Sam since I know he has experience with crypto and hash functions that I don't have).

**Comment by samphippen** [comment]
@myronmarston I'll take a look over this PR this evening and let you know what my thoughts are (if any)

**Comment by samphippen** [comment]
@myronmarston this looks great to me.

**Pull request merged**